### PR TITLE
BuildBot fetcher implementation

### PR DESCRIPTION
**Pull request by holounic**
This PR adds the __BuildBotClient__ that is needed to fetch data from the external BuildBot API and add it to the storage in the view of POJO. 

```BuildBotClient``` class contains ```run``` methods that starts fetching information from the BuildBot with the given name.

```BuildBotPopulation``` class allows to easily create a group of BuildBot fetchers with the common base url via ```populate``` function. The information about every BuildBot is provided in the view of an instance of the ```BuildBotInitializer``` class that contains name, initial build id of a build bot and a request frequency.

For testing, the server mock ```MockWebServer``` is used. This eliminates the possibility that tests will fail due to server errors and allows to specify the server's behaviour and verify that the response JSON is parsed correctly. All JSONs were taken from the BuildBot API.

__+1,306 is only because of two long test json files__ 

Tests cover four different test cases:
- When server response is valid, the client should invoke the ```updateRevisionEntry(arg)``` method of the ```DatastoreRepository``` instance, where the arg is the expected POJO.
- When server responds with an error, the client should not invoke the ```updateRevisionEntry(arg)``` method of the ```DatastoreRepository``` instance.
- When server responds with an empty JSON, the client should not invoke the ```updateRevisionEntry(arg)``` method of the ```DatastoreRepository``` instance.
- When the client have enough time to perform more than one request server responds twice with the valid bodies, the client should invoke the ```updateRevisionEntry(arg)``` method of the ```DatastoreRepository``` instance, where the arg is the expected POJO, at least twice.

### CODE REVIEW

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java
Comment (holounic) [inline src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:206]:
```suggestion
  
```

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (MagoDelBlocco) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:1]: Please add the boilerplate license statement in every new file

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (MagoDelBlocco) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:42]: remove full stops from `@param` sections in JavaDocs. They are needed for the description and `@return` sections only.

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (MagoDelBlocco) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:23]: Consider adding a more in-depth explanation of what the base URL is here. Even though it is explained in the constructor, I believe this should be clear in the field explanation as well.

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClientPopulation.java
Comment (MagoDelBlocco) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClientPopulation.java:7]: This responsibility would have been better delivered using the Factory pattern. Consider changing this.

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java
Comment (MagoDelBlocco) [inline src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:206]: This change doesn't belong in the scope of this PR

**Review by MagoDelBlocco — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClientPopulation.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClientPopulation.java:7]: I don't think the factory pattern would be better than the existing one. I prefer leaving it as it is.

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:23]: Base url is self-explanatory.

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:42]: I looked through other pieces if our code and didn't found anything controversial about adding fullstops at the end of the ```@param``` description

File: src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java
Comment (holounic) [reply src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java:206]:
```suggestion
  
```

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:1]: Done

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (MagoDelBlocco) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:42]: Check the discussion in #23. If you've seen this happen in other parts of our code, please fix those problems as well.

**Review by MagoDelBlocco — APPROVED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:42]: Thanks! Done

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:105]: if `words` contains none of `{"failed", "lost", "successful"}`, the status is not updated. This comment hints that the buildbot is considered `PASSED` in this case

I'd suggest extracting a little helper function that returns the status given the list of `words` similar to the body of this one, that `return`-s where this one `break`-s and that returns `PASSED` by default, when none of the words matched `failed` or `lost`.

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:18]: this import is out of sort order

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:38]: I think it's worth repeating the base URL part of the example from the constructor here. Specifically, consider this extract from the example:
>baseUrl = "http://lab.llvm.org:8011/json/builders"
Someone not familiar with the LLVM BuildBot URLs can reasonably think they can pass any of the following as baseUrl:
* http://lab.llvm.org:8011
* http://lab.llvm.org:8011/json
* http://lab.llvm.org:8011/json/builders

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:83]:
```suggestion
          logger.info(String.format("Error occured, waiting for %d seconds", delay));
```

File: src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTestInfo.java
Comment (krasimir-google) [inline src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTestInfo.java:63]: If `expected.length` is odd and larger than 4 this will fail. Consider detecting this when checking the validity of `expected` and throw an IllegalArgumentException in this case

**Review by krasimir-google — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:66]: I assume this `run` method will be called for different instances concurrently.
To make the logging more useful, consider logging some information to help distinguish between those, such as the `baseUrl`, `buildBot` and the `buildId` that you try fetching.

File: pom.xml
Comment (holounic) [reply pom.xml:119]: The WebClient is a part of it

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:65]: I believe this doesn't have to be atomic. It looks like it's potentially incremented in a separate thread, but the pipeline below doesn't look like it's going to be branching and scheduling multiple concurrent executions of the `subscribe` callback, hence there is no risk of a data race on the value of this variable. Note that I may be missing something here, so if you have concrete reasone where this should necessarily be atomic, please share.

File: src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTest.java
Comment (krasimir-google) [inline src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTest.java:165]: I believe you don't need to call the default `super` constructor; this is done automatically.

File: pom.xml
Comment (krasimir-google) [reply pom.xml:119]: Thank you!

File: src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTest.java
Comment (holounic) [reply src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTest.java:165]: Removed, thanks!

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:65]: It has to be atomic because in Java variable used in lambda expression should be final or effectively final (which means being atomic or final).

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:66]: Done, thanks!

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:17]: Fixed

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:105]: Done, thanks!

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java]: Fixed!

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java]: Done

File: src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTestInfo.java
Comment (holounic) [reply src/test/java/com/google/graphgeckos/dashboard/buildbot/BuildBotClientTestInfo.java:63]: Done

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:106]: I don't see any changes to this comment nor to the `extractStatus` method below. Maybe the PR was not refreshed?

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (krasimir-google) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:65]: Thank you!

**Review by krasimir-google — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (krasimir-google) [inline src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:107]: I think there is a part of this sentence missing.

**Review by krasimir-google — COMMENTED** [review]
Looks good with a comment nit, also in the PR description you could remove the paragraph about `BuildBotPopulation` as that doesn't appear here.

Please wait for a readability review from MForster.

**Comment by krasimir-google** [comment]
LGTM

**Review by krasimir-google — APPROVED** [review]

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (MForster) [inline src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:99]: Same here.

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (MForster) [inline src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:113]: Is it possible that the list is empty?

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (MForster) [inline src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:136]: A list of arrays is an usual data structure. Do you get that from some library?

File: src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java
Comment (MForster) [inline src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:44]: `Preconditions.checkNotNull()` is preferred

Comment (MForster) [reply src/main/java/com/google/graphgeckos/dashboard/fetchers/buildbot/BuildBotClient.java:44]: and then use it directly in the assignment.

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:113]: I haven't seen that in real BuildBot API JSONs, but in this case the BuildBot will be considered as failed.

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:96]: Fixed, thank you!

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:136]: No, I thought that the String[] is an Object so it could be used as a type of a generic. But I still replaced it by the List<List<String>>

File: src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java
Comment (holounic) [reply src/main/java/com/google/graphgeckos/dashboard/datatypes/BuildBotData.java:94]: Fixed

**Pull request merged**
